### PR TITLE
wonderswan updates

### PR DESCRIPTION
**Pull request by wilbertpol**
bus/wswan: Add preliminary support for wonderwitch. [Wilbert Pol]

cpu/v30mz: [Wilbert Pol]
- Rename segment registers.
- Add a callback to retrieve the interrupt vector.
- Add prefetch queue and PFP register.
- Add 1 cycle penalty when reading or writing a word from an unaligned address.
- Add 1 cycle penalty when calculating effective address from multiple registers.
- Removed not supported nec extended instructions.
- Change address space width to word.

wswan: [Wilbert Pol]
- Adjust cpu cycles when performing DMA.
- Updated default internal eeprom for WonderSwan Color with name ‘WONDERSWANCOLOR’.
- Fix off-by-one error in window drawing.
- Major update to the software lists documenting hardware and rom labels and locations.

wscolor - New working software list additions
----------------------------------------------
Digimon Adventure 02 - D1 Tamers (Rev 1)
Digital Monster Card Game - Ver. WonderSwan Color (Rev 2)
Digital Monster - D-Project (Rev 1)
Digimon Tamers - Digimon Medley
Hunter X Hunter - Greed Island (Rev 1)
From TV Animation One Piece - Treasure Wars (Rev 1)

wscolor - New not working software list additions
-------------------------------------------------
mama Mitte'

wscolor - Replaced software list items
--------------------------------------
Dicing Knight. [anon]
Judgement Silversword - Rebirth Edition (Rev. SC21)

wswan - New working software list additions
---------------------------------------------
Chocobo no Fushigi na Dungeon for WonderSwan (Rev 3)
Chocobo no Fushigi na Dungeon for WonderSwan (Rev 1)
Digimon Adventure 02 - Tag Tamers (Rev 0)
Digimon Adventure - Anode Tamer (Rev 1)
Digimon Adventure Campaign Limited Version
Keitai Gyogun Tanchiki - WonderSwan Handy Sonar (Rev 2)
Harobots (Rev 1)
Hunter X Hunter - Ishi o Tsugu Mono (Rev 1)
Mahjong Touryuumon (Rev 1)
Super Robot Taisen Compact (Rev 1)
SD Gundam - Emotional Jam (Rev 2)
SD Gundam Gashapon Senki - Episode 1 (alt)

wscolor - Software list items promoted to working
--------------------------------------------------
Super Robot Taisen Compact for WonderSwan Color
Cardcaptor Sakura - Sakura to Fushigi na Clow Card

wswan - Replaced software list items
-------------------------------------
Armored Unit
Crazy Climber
Clock Tower for WonderSwan
Glocal Hexcite
Gomoku Narabe & Reversi - Touryuumon
Hanafuda Shiyouyo
Keiba Yosou Shien Soft - Yosou Shinkaron
Morita Shougi for WonderSwan
Nobunaga no Yabou for WonderSwan
Sangokushi II for WonderSwan
Shougi Touryuumon
Slither Link
Soccer Yarou! - Challenge the World
Taikyoku Igo - Heisei Kiin
Trump Collection 2 - Bottom-Up Teki Sekaiisshuu no Tabi
Trump Collection - Bottom-Up Teki Trump Seikatsu

wswan - Software list items promoted to working
-------------------------------------------------
Clock Tower for WonderSwan
Meitantei Conan - Nishi no Meitantei Saidai no Kiki!
Super Robot Taisen Compact 2 - Dai-1-bu - Chijou Gekidou Hen
Super Robot Taisen Compact 2 - Dai-2-bu - Uchuu Gekishin Hen (Rev 4)
Super Robot Taisen Compact 2 - Dai-3-bu - Ginga Kessen Hen (Rev 2)
Super Robot Taisen Compact (Rev 0)

wswan - Software list items demoted to not working
----------------------------------------------------
Magical Drop for WonderSwan
MobileWonderGate (Rev 1)
Wonder Borg Robot Works
Robot Works (Asia)

### CODE REVIEW

**Review by cuavas — COMMENTED** [review]
This is one of those nasty cases where the CPU is capable of dynamic bus sizing.  MAME doesn’t really have a neat solution for handling this at the moment.

File: src/mame/audio/wswan.cpp [hunk 1]
```diff
@@ -80,42 +80,19 @@ void wswan_sound_device::device_start()
...
+	{
+		save_item(NAME(m_audio[i].freq), i);
+		save_item(NAME(m_audio[i].period), i);
+		save_item(NAME(m_audio[i].pos), i);
+		save_item(NAME(m_audio[i].vol_left), i);
+		save_item(NAME(m_audio[i].vol_right), i);
+		save_item(NAME(m_audio[i].on), i);
+		save_item(NAME(m_audio[i].offset), i);
+		save_item(NAME(m_audio[i].signal), i);
+	}
```
[the remark this answers is not in the document]
Comment (wilbertpol) [reply src/mame/audio/wswan.cpp]: Changed to use STRUCT_MEMBER

**Comment by wilbertpol** [comment]
> This is one of those nasty cases where the CPU is capable of dynamic bus sizing. MAME doesn’t really have a neat solution for handling this at the moment.

As far as I understood it, the v30mz core itself is using a 16bit bus but the SoCs used in the wonderswans have the possibility to indeed switch the external bus between 8 and 16 bit. The bootroms however always select 16bit and the cartridges are only wired for 16 bit bus.

**Comment by cuavas** [comment]
> The bootroms however always select 16bit and the cartridges are only wired for 16 bit bus.

If only it were that simple…  The cartridge data bus is 16 bits wide, but the cartridge address bus includes the A0 line.  By default, 8-bit accesses are used for 0x010000-0x01ffff and 16-bit accesses are used for 0x020000-0x0fffff.  The A0 line is only used for 8-bit accesses, it’s ignored for 16-bit accesses.  Writes are only emitted for 0x010000-0x01ffff (the SoC black-holes writes to 0x020000-0x0fffff).  Accesses to 0x010000-0x01ffff are three cycle for WoderSwan, or one cycle for WonderSwan Color in colour mode.

**Comment by cuavas** [comment]
Does anyone else have any comments on this?  The bulk of the changes are changing the CPU to use a 16-bit address space.

**Comment by Osso13** [comment]
Since noone has commented in almost two months, last call: Has anyone something to comment? Otherwise in 24 hours I'll merge.

**Comment by Tafoid** [comment]
AFAIK, @wilbertpol was working on this locally first to work out any recent changes before submitting.  Please be sure to get talk to him first before anything.  I can confirm at least the softwarelist updates as shown are valid.

**Comment by wilbertpol** [comment]
I had already updated the PR against latest changes a while ago

**Comment by cuavas** [comment]
I wasn’t saying this can’t be merged when I commented, I was just lamenting the fact that this is a system that MAME doesn’t handle very well right now.

**Comment by cuavas** [comment]
What are the credits for all the new software list items in wswan and wscolor?

**Comment by wilbertpol** [comment]
I got the new ones from redump

**Pull request merged**